pkg/certificateprovider/azurekeyvault/provider.go

```diff
@@ -0,0 +1,207 @@
+		formatKeyVaultCertificate(&keyVaultCert)
+
+		keyVaultCerts = append(keyVaultCerts, keyVaultCert)
+	}
+
+	common.LogDebug("unmarshaled %v key vault objects, keyVaultObjects: %v", len(keyVaultCerts), keyVaultCerts)
+
+	if len(keyVaultCerts) == 0 {
+		return nil, fmt.Errorf("no keyvault certificate configured")
+	}
+
+	common.LogDebug("vaultURI %s", keyvaultUri)
+
+	kvClient, err := initializeKvClient(ctx, azureCloudEnv.KeyVaultEndpoint, tenantID, workloadIdentityClientID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get keyvault client, error: %w", err)
+	}
+
+	certs := []types.Certificate{}
+	for _, keyVaultCert := range keyVaultCerts {
+		common.LogDebug("fetching object from key vault, certName %v,  keyvault %v", keyVaultCert.CertificateName, keyvaultUri)
+
+		// fetch the object from Key Vault
+		result, err := getCertificate(ctx, kvClient, keyvaultUri, keyVaultCert)
+		if err != nil {
+			return nil, err
+		}
+
+		for idx := range result {
+			r := result[idx]
+			objectContent := []byte(r.content)
+
+			cert := types.Certificate{
+				CertificateName: keyVaultCert.CertificateName,
+				Content:         objectContent,
+				Version:         r.version,
+			}
+
+			certs = append(certs, cert)
+			common.LogDebug("added certificates %v to response", cert.CertificateName)
+		}
+	}
+	return certs, nil
+}
+
+// formatKeyVaultCertificate formats the fields in KeyVaultCertificate
+func formatKeyVaultCertificate(object *types.KeyVaultCertificate) {
+	if object == nil {
+		return
+	}
+	objectPtr := reflect.ValueOf(object)
+	objectValue := objectPtr.Elem()
+
+	for i := 0; i < objectValue.NumField(); i++ {
+		field := objectValue.Field(i)
+		if field.Type() != reflect.TypeOf("") {
+			continue
+		}
+		str := field.Interface().(string)
+		str = strings.TrimSpace(str)
+		field.SetString(str)
+	}
+}
+
+// parseAzureEnvironment returns azure environment by name
+func parseAzureEnvironment(cloudName string) (*azure.Environment, error) {
+	var env azure.Environment
+	var err error
+	if cloudName == "" {
+		env = azure.PublicCloud
+	} else {
+		env, err = azure.EnvironmentFromName(cloudName)
+	}
+	return &env, err
+}
+
+func initializeKvClient(ctx context.Context, keyVaultEndpoint, tenantID, clientId string) (*kv.BaseClient, error) {
+	kvClient := kv.New()
+	kvEndpoint := strings.TrimSuffix(keyVaultEndpoint, "/")
+
+	err := kvClient.AddToUserAgent("ratify")
+	if err != nil {
+		return nil, fmt.Errorf("failed to add user agent to keyvault client, error: %w", err)
+	}
+
+	kvClient.Authorizer, err = getAuthorizerForWorkloadIdentity(ctx, tenantID, clientId, kvEndpoint)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get authorizer for keyvault client, error: %w", err)
+	}
+	return &kvClient, nil
+}
+
+// getCertificate retrieves the certificate from the vault
+func getCertificate(ctx context.Context, kvClient *kv.BaseClient, vaultURL string, kvObject types.KeyVaultCertificate) ([]keyvaultObject, error) {
```

Comment:
```suggestion
func getCertificate(ctx context.Context, kvClient *kv.BaseClient, vaultURL string, kvObject types.KeyVaultCertificate) (keyvaultObject, error) {
```

This function is only returning a single cert, so returning a slice isn't required